### ISSUE TITLE
find_learning_rate step problem

### ISSUE THREAD

**Issue by BernierCR** [zalandoresearch/flair#986]
**Describe the bug**

find_learning_rate does not work. It will use the learning rate from iteration 2 for all iterations after 2. 

**To Reproduce**

This will happen with any application of find_learning_rate. You can reproduce with the tutorial code if you like or any code. 

**Expected behavior**
I expect iteration 3 to use the learning rate for iteration 3, and 4 for 4, and so on. It should end at the maximum learning rate set. 

**Screenshots**

**Environment (please complete the following information):**
 - OS: Linux of course
 - Version MASTER

**Additional context**

scheduler = ExpAnnealLR(optimizer, end_learning_rate, iterations)
scheduler.step(1)
learning_rate = scheduler.get_lr()[0]

def get_lr(self):
        iteration = self.last_epoch + 1
        pct = iteration / self.iterations
        return [base_lr * (self.end_lr / base_lr) ** pct for base_lr in self.base_lrs]

So get_lr is iterating based on last_epoch. Though it's really supposed to be iterating based on mini-batches. Anyways, it gets worse. 

ExpAnnealLR does not have it's own step method. So it uses the base class one. 

def step(self, epoch=None):
        if epoch is None:
            epoch = self.last_epoch + 1
        self.last_epoch = epoch
        for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
            param_group['lr'] = lr

This function expects the correct epoch number to be passed to it.  Fine. 

But it's used like this in this case. 

scheduler.step(1)

This makes the epoch always come out as two. 

so either 
1) find_learning_rate needs to pass the correct iteration number to step
or,
2) the ExpAnnealLR  class needs it's own step function to overide the default, that will correctly increment itself. 

Considering nothing else uses ExpAnnealLR, I'm going with solution 2 in my copy. You do what you like in your package.

**Comment by BernierCR** [comment]
Also, in class _LRScheduler ,

last_epoch is not saved. It should be saved to self.last_epoch

**Comment by BernierCR** [comment]
I also changed the logic to use a repeating iterator instead of the oneshot kind, to loop through batches. People with a low number of mini_batches relative to training set size, need more then one epochs worth of iterations, so it needs to go back to the beginning and keep going until it diverges, or max iterations is reached.

**Comment by alanakbik** [comment]
@BernierCR thanks for sharing these findings. I've added this to the learning rate finder, i.e. iterating over multiple epochs if necessary and increasing the step manually.
